### PR TITLE
New report

### PR DESCRIPTION
**Pull request by prateksha**
# Description
- Admin column added to the User table. User can been made an admin via the rails console by maintainers.
- Report model includes 
  - reporter_id
  - reportee_id
  - comments
- 'Report' button displayed for each ally in the allies list. On clicking 'Report' button, User is redirected to a form for additional comments.  
- Report dashboard can be viewed by the admin users. Reports can be deleted by the admin user after reviewing. Link to the Admin panel has been included in the header for admin users.

## Corresponding Issue
Worked on the following tickets related to [Issue #14](https://github.com/ifmeorg/ifme/issues/14):
- [Create admin User privileges - #901](https://github.com/ifmeorg/ifme/issues/901)
- [Report Model - #863](https://github.com/ifmeorg/ifme/issues/863)
- [Report button on User profiles - #870](https://github.com/ifmeorg/ifme/issues/870)
- [Report Dashboard for admin - #861](https://github.com/ifmeorg/ifme/issues/861)

# Test Coverage
Working on tests for report model.

# Work In Progress
Working on following issues :
- [ ] Create modal for report form

- [ ]  Mail user who reported the other user

- [ ]  Include i18n

- [ ] Write tests

### CODE REVIEW

**Comment by julianguyen** [comment]
I set up a staging branch for work that is being incrementally done and can't be merged to master right away. So instead of merging to master for this reporting feature project, let's merge to staging. Once everything is completed we'll merge to master.

File: Gemfile.lock [hunk 1]
```diff
@@ -96,6 +96,13 @@ GEM
     cocaine (0.6.0)
       terrapin (= 0.6.0)
     coderay (1.1.2)
+    coffee-rails (4.2.2)
```
Comment (julianguyen) [inline Gemfile.lock]: Ideally, we shouldn't be using coffescript in our dependencies. Looks like this is coming from rails_admin :/ We using a lot of dependencies just to be able to view insights into reports. 

I really recommend writing our own admin dashboard, that way we don't have to bring in more unnecessary dependencies. All we have to do is set up a feature flag for admin users in Devise and then build pages where only those users can view data.

Comment (julianguyen) [reply Gemfile.lock]: Devise has official documentation for this: https://github.com/plataformatec/devise/wiki/How-To:-Add-an-Admin-Role

File: app/assets/javascripts/reports.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (julianguyen) [inline app/assets/javascripts/reports.js]: I would only create this file once there are things inside of it.

File: app/assets/stylesheets/reports.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Reports controller here.
```
Comment (julianguyen) [inline app/assets/stylesheets/reports.scss]: I would only create this file once there are things inside of it.

File: spec/models/report_spec.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (julianguyen) [inline spec/models/report_spec.rb]: I would only create this file once there are things inside of it.

File: spec/helpers/reports_helper_spec.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
```
Comment (julianguyen) [inline spec/helpers/reports_helper_spec.rb]: I would only create this file once there are things inside of it.

File: spec/controllers/reports_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (julianguyen) [inline spec/controllers/reports_controller_spec.rb]: Please write tests for this

File: db/migrate/20180730034853_add_admin_to_user.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class AddAdminToUser < ActiveRecord::Migration[5.0]
+  def change
+    add_column :users, :admin, :boolean
```
Comment (julianguyen) [inline db/migrate/20180730034853_add_admin_to_user.rb]: This LGTM

File: app/controllers/reports_controller.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+class ReportsController < ApplicationController
+  def new
```
Comment (julianguyen) [inline app/controllers/reports_controller.rb]: Remove this if unused

File: app/helpers/reports_helper.rb [hunk 9]
```diff
@@ -0,0 +1,2 @@
+module ReportsHelper
```
Comment (julianguyen) [inline app/helpers/reports_helper.rb]: I would only create this file once there are things inside of it.

File: app/controllers/reports_controller.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+  end
+
+  def create
+    user_id = current_user.id
+    ally_id = params[:ally_id].to_i
+    @report = Report.create(reporter_id: user_id, reportee_id: ally_id, comments: params[:report][:comments] )
+  #  ReportMailer.reporter_email(@user).deliver
```
Comment (julianguyen) [inline app/controllers/reports_controller.rb]: In a comment, I would specify why this is commented out for now.

File: db/schema.rb [hunk 11]
```diff
@@ -180,6 +180,14 @@
     t.datetime "updated_at"
   end
 
+  create_table "reports", force: :cascade do |t|
```
Comment (julianguyen) [inline db/schema.rb]: Where is the corresponding migration file for this?

**Comment by aSquare14** [comment]
I'll migrate it to devise !

**Comment by prateksha** [comment]
@julianguyen @camillevilla @alvarocasadoc . 

- We have used Devise for Admin dashboard.
- Report feature has been implemented for the following:
    - User Profile
    - Allies List,
    - Comments in ( Moments, Strategies, Meetings)
- Report Table contains reporter_id, reportee_id, description, comment_type, comment_id(if applicable)

Please review the PR.

File: app/controllers/reports_controller.rb [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    user_mail = current_user.email
+    comment_id = params[:comment_id]
+    commentable_type = params[:commentable_type]
+    if comment_id == nil
+      comment_id = -1
+    end
+    if commentable_type == nil
+      commentable_type = "user"
+    end
+    @report = Report.create(reporter_id: user_id, reportee_id: ally_id, comments: params[:report][:comments], commentable_id: comment_id, commentable_type: commentable_type )
```
Comment (camillevilla) [inline app/controllers/reports_controller.rb]: This line is getting pretty long. To keep this readable, how about breaking these attributes onto several lines? For example:
```ruby
@report = Report.create(
  reporter_id: user_id,
  reportee_id: ally_id
  ...
```

**Pull request closed without merging**